ProjFS.Mac/PrjFSKextTests/HandleOperationTests.mm

```diff
@@ -706,5 +712,176 @@ - (void) testCloseWithoutModifed {
+    HandleVnodeOperation(
+        nullptr,
+        nullptr,
+        KAUTH_VNODE_DELETE,
+        reinterpret_cast<uintptr_t>(context),
+        reinterpret_cast<uintptr_t>(testFileVnode.get()),
+        0,
+        0);
+
+    XCTAssertFalse(MockCalls::DidCallFunction(ProviderMessaging_TrySendRequestAndWaitForResponse));
+}
+
+- (void) testDeleteWhenRequestFails {
+    ProviderMessageMock_SetDefaultRequestResult(false);
+    testFileVnode->attrValues.va_flags = FileFlags_IsEmpty | FileFlags_IsInVirtualizationRoot;
+    HandleVnodeOperation(
+        nullptr,
+        nullptr,
+        KAUTH_VNODE_DELETE,
+        reinterpret_cast<uintptr_t>(context),
+        reinterpret_cast<uintptr_t>(testFileVnode.get()),
+        0,
+        0);
+
+   XCTAssertTrue(MockCalls::CallCount(ProviderMessaging_TrySendRequestAndWaitForResponse) == 1);
+}
+
+- (void) testDeleteDirectoryWithNoVirtualizationRoot {
+    ProviderMessageMock_SetCleanupRootsAfterRequest(true);
+    testDirVnode->attrValues.va_flags = FileFlags_IsInVirtualizationRoot;
+    HandleVnodeOperation(
+        nullptr,
+        nullptr,
+        KAUTH_VNODE_DELETE,
+        reinterpret_cast<uintptr_t>(context),
+        reinterpret_cast<uintptr_t>(testDirVnode.get()),
+        0,
+        0);
+
+    XCTAssertTrue(MockCalls::CallCount(ProviderMessaging_TrySendRequestAndWaitForResponse) == 1);
+}
+
+
+- (void) testDeleteDirectoryWhenSecondRequestFails {
+    ProviderMessageMock_SetSecondRequestResult(false);
+    testDirVnode->attrValues.va_flags = FileFlags_IsInVirtualizationRoot;
+    HandleVnodeOperation(
+        nullptr,
+        nullptr,
+        KAUTH_VNODE_DELETE,
+        reinterpret_cast<uintptr_t>(context),
+        reinterpret_cast<uintptr_t>(testDirVnode.get()),
+        0,
+        0);
+
+    XCTAssertTrue(MockCalls::CallCount(ProviderMessaging_TrySendRequestAndWaitForResponse) == 2);
+}
+
+- (void) testReadAttributesDirectoryWithNoVirtualizationRoot {
+    VirtualizationRoots_Cleanup();
+    testDirVnode->attrValues.va_flags = FileFlags_IsEmpty | FileFlags_IsInVirtualizationRoot;
+    HandleVnodeOperation(
+        nullptr,
+        nullptr,
+        KAUTH_VNODE_READ_ATTRIBUTES,
+        reinterpret_cast<uintptr_t>(context),
+        reinterpret_cast<uintptr_t>(testDirVnode.get()),
+        0,
+        0);
+
+    XCTAssertFalse(MockCalls::DidCallFunction(ProviderMessaging_TrySendRequestAndWaitForResponse));
+}
+
+- (void) testReadAttributesDirectoryWhenRequestFails {
+    ProviderMessageMock_SetDefaultRequestResult(false);
+    testDirVnode->attrValues.va_flags = FileFlags_IsEmpty | FileFlags_IsInVirtualizationRoot;
+    HandleVnodeOperation(
+        nullptr,
+        nullptr,
+        KAUTH_VNODE_READ_ATTRIBUTES,
+        reinterpret_cast<uintptr_t>(context),
+        reinterpret_cast<uintptr_t>(testDirVnode.get()),
+        0,
+        0);
+
+    XCTAssertTrue(MockCalls::CallCount(ProviderMessaging_TrySendRequestAndWaitForResponse) == 1);
```

Comment:
Is there anything else we can/should validate happens when a request fails? (e.g. should the kext be logging an error?)

This comment applies to all of the tests for the failure scenario.

Reply:
I was thinking that too.  It can validly fail because it's shutting down (probably don't want to log that). Another case would be if the userClient became null.  In `ActiverProvider_SendMessage` we kernal panic if the virtualization root is invalid.  

Perhaps it wouldn't hurt to add a call saying the send failed, just curious if we thought it would be extra noise.  It does seem like this would drop a message.

Reply:
> Is there anything else we can/should validate happens when a request fails? 

> I was thinking that too

In my earlier pass over the changes I missed that we're not validating that `HandleVnodeOperation` is returning the result we expect it to:


_From HandleVnodeOperation_
```
CleanupAndReturn:
    if (putVnodeWhenDone)
    {
        vnode_put(currentVnode);
    }
    
    atomic_fetch_sub(&s_numActiveKauthEvents, 1);
    return kauthResult;
```

Can you add that validation to the `HandleVnodeOperation` tests?  That will allow us to verify that `HandleVnodeOperation` is properly passing through the result returned by `ProviderMessaging_TrySendRequestAndWaitForResponse`.

(It looks like the `ProviderMessaging_TrySendRequestAndWaitForResponse` mock will need to be updated to set `kauthResult` and `kauthError` properly)

Reply:
Testing kauthResult/kauthError for `TryGetVirtualizationRoot ` is straightforward and since we are calling the actual function there we can easily validate.

Failures of `ProviderMessaging_TrySendRequestAndWaitForResponse ` are a bit more hairy.  In the mock I can make the fail cases return an error and we can validate on that.  

In the actual function shutting down for instance on `ProviderMessaging_TrySendRequestAndWaitForResponse ` would return false and kauthResult/kauthError would be unchanged.  Perhaps these should be set?

Reply:
> Testing kauthResult/kauthError for TryGetVirtualizationRoot is straightforward and since we are calling the actual function there we can easily validate.

That sounds good.

> Failures of ProviderMessaging_TrySendRequestAndWaitForResponse are a bit more hairy. In the mock I can make the fail cases return an error and we can validate on that.

That sounds good to me too, and if you think it will be a major change I'm good with spinning out a follow up issue/PR for enhancing the `ProviderMessaging` mock and our validation of `ProviderMessaging` failure scenarios.

> In the actual function shutting down for instance on ProviderMessaging_TrySendRequestAndWaitForResponse would return false and kauthResult/kauthError would be unchanged. Perhaps these should be set?

I agree it would be best to set kauthResult/kauthError in that scenario.  Fixing up `ProviderMessaging_TrySendRequestAndWaitForResponse` seems outside of the scope of this PR though and so I think that work could be deferred to a follow up issue.